### PR TITLE
Enable remote registry proxy everwhere

### PR DESCRIPTION
**Pull request by mikejholly**
Note: `EARTHLY_GLOBAL_WAIT_END=true` also fixes the issue without this hacky patch. Should we just wait until global wait block is enabled?

### CODE REVIEW

File: builder/builder.go [hunk 1]
```diff
@@ -217,11 +217,19 @@ func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt
 		// to accomodate parallelism in the WAIT/END PopWaitBlock handling
 		dirIDs = map[int]string{}
 	)
+
```
Comment (alexcb) [inline builder/builder.go:220]: nit: why not just combine the two `var` sections instead?

File: builder/builder.go [hunk 2]
```diff
@@ -217,11 +217,19 @@ func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt
...
 		depIndex   = 0
 		imageIndex = 0
 		dirIndex   = 0
 	)
+
+	// Delay closing the registry proxy server until this method
+	// returns. Previously, it was being closed when the buildFunc returned,
+	// which led to a problem with fetching some images.
+	stopRegistryProxyFunc := func() {}
+	defer stopRegistryProxyFunc()
```
Comment (alexcb) [inline builder/builder.go:231]: I don't think this works, see the following for an example:

```
alex@surfperch:~$ cat main.go 
package main

import "fmt"

func main() {

	stopRegistryProxyFunc := func() {
		fmt.Println("first func")
	}
	defer stopRegistryProxyFunc()

	stopRegistryProxyFunc = func() {
		fmt.Println("is this actually called?")
	}
}
alex@surfperch:~$ go run main.go 
first func
```

Comment (mikejholly) [reply builder/builder.go:231]: Good point. I was originally using the below approach but wanted to condense the code. It works, but this is because the registry never closes.

```golang
	var stopRegistryProxyFunc func()
	defer func() {
		if stopRegistryProxyFunc != nil {
			stopRegistryProxyFunc()
		}
	}()
```

Comment (alexcb) [reply builder/builder.go:231]: `defer func() { stopRegistryProxyFunc() }` would probably work.

Comment (mikejholly) [reply builder/builder.go:231]: Sure, that's fine too.

**Comment by alexcb** [comment]
we should validate that this fixes the issues that #3624 surfaced prior to merging this.

File: builder/builder.go (hunk 1 above)
Comment (mikejholly) [reply builder/builder.go:220]: I think it's ok to group var & const declarations by purpose or subject.

**Comment by mikejholly** [comment]
> we should validate that this fixes the issues that #3624 surfaced prior to merging this.

Does it make more sense to merge this PR and then address any other issues in a PR to enable remote registry for version 0.8?

**Comment by alexcb** [comment]
> > we should validate that this fixes the issues that #3624 surfaced prior to merging this.
> 
> Does it make more sense to merge this PR and then address any other issues in a PR to enable remote registry for version 0.8?

We should try to add tests to every PR when possible -- that could be done by cherry-picking #3624 and combining them into a single PR, or possibly by creating a new test that enables both `--use-inline-cache --use-remote-registry` features.

**Comment by alexcb** [comment]
Linking #3628 to this PR

File: cmd/earthly/flag/global.go [hunk 3]
```diff
@@ -517,12 +516,5 @@ func (global *Global) RootFlags(installName string, bkImage string) []cli.Flag {
 			Usage:       "A remote docker image tag use as explicit cache",
 			Destination: &global.RemoteCache,
 		},
-		&cli.BoolFlag{
-			Name:        "use-remote-registry",
-			EnvVars:     []string{"EARTHLY_USE_REMOTE_REGISTRY"},
-			Usage:       "Use remote registry with Docker pull",
-			Destination: &global.UseRemoteRegistry,
-			Hidden:      true, // Internal.
-		},
```
Comment (vladaionescu) [inline cmd/earthly/flag/global.go:526]: I wonder if we should keep this flag around (and just flipping the default value), just in case a user needs to turn it off for some reason (e.g. running into a bug)?

File: builder/builder.go [hunk 4]
```diff
@@ -217,11 +211,21 @@ func (b *Builder) convertAndBuild(ctx context.Context, target domain.Target, opt
...
 	)
+
 	var (
 		depIndex   = 0
 		imageIndex = 0
 		dirIndex   = 0
 	)
+
+	// Delay closing the registry proxy server until this method
+	// returns. Previously, it was being closed when the buildFunc returned,
```
Comment (alexcb) [inline builder/builder.go:222]: nit: no need to mention a past bug that was fixed in this comment.

I would reword this as simply: `Delay closing the registry proxy server until after the buildFunc returns`

File: cmd/earthly/flag/global.go (hunk 3 above)
Comment (alexcb) [reply cmd/earthly/flag/global.go:526]: I'm on the fence regarding this, yes it's a nice escape hatch; however keeping around all these old features adds to maintenance cost.

We could leave it as-is for a few releases, and then fully remove it down the road; alternatively if a user runs into the need for this, we could re-introduce it, and the user could pin to an older version until it's done.

Comment (vladaionescu) [reply cmd/earthly/flag/global.go:526]: Yeah, that's reasonable. Let's nuke it then

Comment (mikejholly) [reply cmd/earthly/flag/global.go:526]: @alexcb and I were talking about this as well. On the one hand, it could be useful to retain the option to disable it. On the other, there's a small, accumulating cost with keeping these around (code, tests, etc.).

I'd vote for removing it and fixing bugs as they come up. All the tests are green :green_apple:

File: builder/builder.go (hunk 4 above)
Comment (mikejholly) [reply builder/builder.go:222]: Sure. I wanted to allude to the fact that this code can be reverted once global wait block support is enabled. Updated.

**Review by alexcb — APPROVED** [review]
